### PR TITLE
Add MDETR losses

### PR DESCRIPTION
**Pull request by ebsmothers**
Added the classification and box losses (L1 and GIoU) from MDETR.
Still to add the contrastive alignment loss, will update with that later on.

Test plan:
```
$ python -m pytest -v test/modules/losses/test_mdetr_losses.py
=================================================== test session starts ====================================================
platform linux -- Python 3.8.13, pytest-7.1.2, pluggy-1.0.0 -- /data/home/ebs/miniconda3/envs/torch-multimodal/bin/python
cachedir: .pytest_cache
rootdir: /data/home/ebs/torchmultimodal
collected 2 items

test/modules/losses/test_mdetr_losses.py::TestMDETRLosses::test_classification_loss PASSED                           [ 50%]
test/modules/losses/test_mdetr_losses.py::TestMDETRLosses::test_box_losses PASSED                                    [100%]

==================================================== 2 passed in 3.62s =====================================================
```

Summary:

Test plan:

Fixes #{issue number}

### CODE REVIEW

File: torchmultimodal/modules/losses/mdetr.py [hunk 1]
```diff
@@ -0,0 +1,119 @@
...
+        tgt_idx.append(tgt + offset)
+        offset += n_target_boxes[i]
+    tgt_idx = torch.cat(tgt_idx)
+    tgt_pos = positive_map[tgt_idx]
+    target_sim = torch.zeros_like(logits)
+    target_sim[:, :, -1] = 1
+    target_sim[src_idx] = tgt_pos
+    loss_ce = -(logits * target_sim).sum(-1)
+
+    eos_tensor = torch.full(loss_ce.shape, eos_coef, device=target_sim.device)
```
[the remark this answers is not in the document]
Comment (ebsmothers) [reply torchmultimodal/modules/losses/mdetr.py:71]: This is downweighting the losses on those boxes that are not matched. Kinda related to your other comment but lemme try to make this clearer.

File: torchmultimodal/modules/losses/mdetr.py [hunk 2]
```diff
@@ -0,0 +1,119 @@
...
+    offset = 0
+    for i, (_, tgt) in enumerate(indices):
+        tgt_idx.append(tgt + offset)
+        offset += n_target_boxes[i]
+    tgt_idx = torch.cat(tgt_idx)
+    tgt_pos = positive_map[tgt_idx]
+    target_sim = torch.zeros_like(logits)
+    target_sim[:, :, -1] = 1
+    target_sim[src_idx] = tgt_pos
+    loss_ce = -(logits * target_sim).sum(-1)
```
[the remark this answers is not in the document]
Comment (ebsmothers) [reply torchmultimodal/modules/losses/mdetr.py:69]: It's not necessarily that there are that many extra calculations, but there is a lot of wrangling various indices that needs to occur here. I opted to put this in the loss for now, but we could also consider refactoring some of it into the training loop later on. The advantage of doing it this way is that it is pretty well set up to run immediately after the matcher though

**Review by RdoubleA — APPROVED** [review]
This is a lot more clear now, thanks for making these changes. Looks good to me!

File: test/modules/losses/test_mdetr_losses.py [hunk 3]
```diff
@@ -0,0 +1,118 @@
...
+                pred_logits, n_boxes_per_sample, positive_map, indices, num_boxes
+            )
+        )
+        expected = torch.tensor(5.0893)
+        assert_expected(actual, expected, rtol=0, atol=1e-3)
+
+    def test_box_losses(self, pred_boxes, target_boxes, indices, num_boxes):
+        actual = box_losses(pred_boxes, target_boxes, indices, num_boxes)
+        expected_l1_loss = torch.tensor(0.7721)
+        expected_giou_loss = torch.tensor(1.1768)
```
Comment (CaoE) [inline test/modules/losses/test_mdetr_losses.py:116]: Hi @ebsmothers  May I know why fixed expected losses hard coded here. When adding new implementation for exponential (which will be called by multinomial and change the value of indices) , the fixed `expected_l1_loss = torch.tensor(0.7721)` and `expected_giou_loss = torch.tensor(1.1768)` is not applicable for new exponential.https://github.com/pytorch/pytorch/pull/69967

**Comment by CaoE** [comment]
@ebsmothers  I'm not familiar with MDETR losses, and I see that `pred_boxes` and `target_boxes` are generated randomly, and indices is generated by multinomial.  May I know what is the purpose of `test_box_losses` and why `torch.tensor(0.7721)` and `torch.tensor(1.1768)` are used here.  It seems the actual loss is depend on the value of multinomial. Even the new mkl implementation fits the exponential distribution but it is not the same as the old one.  The change of `indices` generated by multinomial will also change the actual loss. Can we change the expected loss to accommodate the new exponential or use a more flexible test instead of fixed losses ? Thank you very much for any clarification.

File: test/modules/losses/test_mdetr_losses.py (hunk 3 above)
Comment (ebsmothers) [reply test/modules/losses/test_mdetr_losses.py:116]: @CaoE we have validated these fixed values against those from the original model, so by using these the goal is to ensure that the performance remains the same. Can you clarify what you mean by this? 

> the fixed expected_l1_loss = torch.tensor(0.7721) and expected_giou_loss = torch.tensor(1.1768) is not applicable for new exponential

**Comment by CaoE** [comment]
@ebsmothers  PyTorch already uses scipy.stats to test sampling results for exponential, and also tests the var and mean. The new exponential implementation conforms to the exponential distribution. But the new implementation is not the same as the old one. Since multinomial is depend on exponential, multinomial will also be affected, but it will also fit the multinomial distribution and remain deterministic if setting the seed.

For the test case `test_box_losses`, I tried to change the initial value of `set_rng_seed(1)` , `batch_size`, or `num_queries`, `box_losses` also changed a lot accordingly.  Can we relax the test, for example set rtol=5e-2 and atol=5e-2 ?

**Comment by ebsmothers** [comment]
@CaoE thanks for the additional info. I took a pass at #364, happy to continue the discussion there.
